clang/lib/3C/CastPlacement.cpp

```diff
@@ -18,107 +18,244 @@
+  // WILD on top of a wild expression.
+  if (SrcInt->isChecked(E) && (UncheckedItypeCall || FptrCall) &&
+      !DstExt->solutionEqualTo(Info.getConstraints(), SrcInt))
+    return CastNeeded::CAST_TO_WILD;
+
+  // If nothing above returned, then no casting is required.
+  return CastNeeded::NO_CAST;
 }
 
-// Get the type name to insert for casting.
-std::string CastPlacementVisitor::getCastString(ConstraintVariable *Src,
-                                                ConstraintVariable *Dst) {
-  assert(needCasting(Src, Dst) && "No casting needed.");
-  return "(" + Dst->getRewritableOriginalTy() + ")";
+// Get the string representation of the cast required  at all call. The return
+// is a pair of strings: a prefix and suffix string that for the complete cast
+// when placed around the expression being cast.
+std::pair<std::string, std::string>
+CastPlacementVisitor::getCastString(const ConstraintVariable *SrcInt,
+                                    const ConstraintVariable *SrcExt,
+                                    const ConstraintVariable *DstInt,
+                                    const ConstraintVariable *DstExt) {
+  CastNeeded CastType = needCasting(SrcInt, SrcExt, DstInt, DstExt);
+  const auto &E = Info.getConstraints().getVariables();
+  switch (CastType) {
+    case CAST_TO_WILD:
+      return std::make_pair("((" + DstExt->getRewritableOriginalTy() + ")",
+                            ")");
+    case CAST_TO_CHECKED: {
+      std::string Suffix = ")";
+      if (const auto *DstPVC = dyn_cast<PVConstraint>(DstExt)) {
+        assert("Checked cast not to a pointer" && !DstPVC->getCvars().empty());
+        ConstAtom *CA = Info.getConstraints().getAssignment(
+          DstPVC->getCvars().at(0));
+
+        // Writing an _Assume_bounds_cast to an array type requires inserting
+        // the bounds for destination array. These can come from the source
+        // code or the infered bounds. If neither source is available, use empty
+        // bounds.
+        if (isa<ArrAtom>(CA) || isa<NTArrAtom>(CA)) {
+          std::string Bounds = "";
+          if (DstPVC->srcHasBounds())
+            Bounds = DstPVC->getBoundsStr();
+          else if (DstPVC->hasBoundsKey())
+            Bounds = ABRewriter.getBoundsString(DstPVC, nullptr, true);
+          if (Bounds.empty())
+            Bounds = "byte_count(0)";
+
+          Suffix = ", " + Bounds + ")";
+        }
+      }
+      return std::make_pair(
+        "_Assume_bounds_cast<" + DstExt->mkString(E, false) + ">(", Suffix);
+    }
+    default:
+      llvm_unreachable("No casting needed");
+  }
 }
 
-void CastPlacementVisitor::surroundByCast(const std::string &CastPrefix,
+
+void CastPlacementVisitor::surroundByCast(const ConstraintVariable *SrcInt,
+                                          const ConstraintVariable *SrcExt,
+                                          const ConstraintVariable *DstInt,
+                                          const ConstraintVariable *DstExt,
```

Comment:
To confirm: If the cast can't happen for whatever reason, it's not added, so the programmer has to do it?

Reply:
Yes, sometimes it's not possible to add the cast (primarily due to macros). I might like to change constraint generation so that it will never need to add a cast inside a macro (this would mean adding equality constraints that make more pointers wild), or maybe we're happy requiring the programmer to make some manual fixes.

Reply:
Sure, that's fine until we have a better plan.